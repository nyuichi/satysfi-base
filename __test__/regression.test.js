const fs = require("fs");
const shell = require("shelljs");
const tmp = require("tmp");
const { toMatchPdfSnapshot } = require("jest-pdf-snapshot");
expect.extend({ toMatchPdfSnapshot });

shell.config.fatal = true;

function compileSatysfi(src) {
  const tmpFile = tmp.fileSync();

  const { code: exitCode } = shell.exec(`satysfi ${src} -o ${tmpFile.name}`, {
    silent: true,
  });

  const pdfBuffer = fs.readFileSync(tmpFile.name);
  tmpFile.removeCallback();

  return {
    exitCode,
    pdfBuffer
  };
}

shell.cd("__test__");

afterAll(() => {
  shell.rm("-f", "test.pdf", "satysrc/**/*.satysfi-aux");
});

test("Confirm that satysfi is installed", () => {
  expect(shell.exec("satysfi --version").code).toBe(0);
});

test("Check compiler outputs", () => {
  const compilerOutput = shell
    .exec("satysfi satysrc/generic.saty", { silent: true })
    .exec(
      "awk '/evaluating texts .../{flag=1;next}/evaluation done/{flag=0}flag'",
      { silent: true }
    ).stdout;
  expect(compilerOutput).toMatchSnapshot();
});

test("SATySFi-iT", () => {
  const result = compileSatysfi("satysrc/satysfi-it.saty");

  expect(result.exitCode).toBe(0);
  expect(result.pdfBuffer).toMatchPdfSnapshot();
});

describe("Derive", () => {
  const filenames = [
    "01-derive",
    "02-assume",
    "03-by",
    "04-by-and-byop",
    "05-from",
    "06-dotted-line",
    "07-customized-config",
  ];

  for (const filename of filenames) {
    test(`renders ${filename}`, () => {
      const result = compileSatysfi(`satysrc/derive/${filename}.saty`);

      expect(result.exitCode).toBe(0);
      expect(result.pdfBuffer).toMatchPdfSnapshot();
    });
  }
});
