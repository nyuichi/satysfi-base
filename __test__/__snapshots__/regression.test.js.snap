// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check compiler outputs 1`] = `
"[32m[debug.log] ==== Ref ====[0m
[32m[debug.log] 1[0m
[32m[debug.log] 2[0m
[32m[debug.log] 2:1[0m
[32m[debug.log] 2[0m
[32m[debug.log] 1[0m
[32m[debug.log] ==== String ====[0m
[32m[debug.log] abc[0m
[32m[debug.log] true[0m
[32m[debug.log] spam[0m
[32m[debug.log] ham[0m
[32m[debug.log] eggs[0m
[32m[debug.log] [0m
[32m[debug.log] a[0m
[32m[debug.log] b[0m
[32m[debug.log] c[0m
[32m[debug.log] true[0m
[32m[debug.log] true[0m
[32m[debug.log] true[0m
[32m[debug.log] true[0m
[32m[debug.log] ==== List ====[0m
[32m[debug.log] 12345[0m
[32m[debug.log] 12345[0m
[32m[debug.log] 12345[0m
[32m[debug.log] (1,2)(1,1)(2,1)[0m
[32m[debug.log] (1,2)(1,1)(2,1)[0m
[32m[debug.log] (1,2)(1,1)(2,1)[0m
[32m[debug.log] ==== Array ====[0m
[32m[debug.log] 3[0m
[32m[debug.log] 43[0m
[32m[debug.log] 0[0m
[32m[debug.log] 1[0m
[32m[debug.log] 2[0m
[32m[debug.log] 3[0m
[32m[debug.log] 4[0m
[32m[debug.log] 5[0m
[32m[debug.log] 6[0m
[32m[debug.log] 7[0m
[32m[debug.log] 8[0m
[32m[debug.log] 9[0m
[32m[debug.log] 10[0m
[32m[debug.log] 9[0m
[32m[debug.log] 8[0m
[32m[debug.log] 7[0m
[32m[debug.log] 6[0m
[32m[debug.log] 5[0m
[32m[debug.log] 4[0m
[32m[debug.log] 3[0m
[32m[debug.log] 2[0m
[32m[debug.log] 1[0m
[32m[debug.log] 10[0m
[32m[debug.log] 9[0m
[32m[debug.log] 8[0m
[32m[debug.log] 7[0m
[32m[debug.log] 6[0m
[32m[debug.log] 5[0m
[32m[debug.log] 4[0m
[32m[debug.log] 3[0m
[32m[debug.log] 2[0m
[32m[debug.log] 1[0m
[32m[debug.log] 0[0m
[32m[debug.log] 0[0m
[32m[debug.log] 1[0m
[32m[debug.log] 2[0m
[32m[debug.log] 3[0m
[32m[debug.log] 4[0m
[32m[debug.log] 5[0m
[32m[debug.log] 6[0m
[32m[debug.log] 7[0m
[32m[debug.log] 8[0m
[32m[debug.log] 9[0m
[32m[debug.log] ==== RegExp ====[0m
[32m[debug.log] true[0m
[32m[debug.log] 0: SPLIT 1 3[0m
[32m[debug.log] 1: CHAR a[0m
[32m[debug.log] 2: JMP 2[0m
[32m[debug.log] 3: CHAR b[0m
[32m[debug.log] 4: MATCH[0m
[32m[debug.log] true[0m
[32m[debug.log] 0: SAVE 0[0m
[32m[debug.log] 1: SPLIT 1 4[0m
[32m[debug.log] 2: CHAR a[0m
[32m[debug.log] 3: SPLIT -1 1[0m
[32m[debug.log] 4: JMP 4[0m
[32m[debug.log] 5: SAVE 2[0m
[32m[debug.log] 6: CHAR c[0m
[32m[debug.log] 7: SAVE 3[0m
[32m[debug.log] 8: SAVE 1[0m
[32m[debug.log] 9: SAVE 4[0m
[32m[debug.log] 10: CHAR b[0m
[32m[debug.log] 11: SPLIT -1 1[0m
[32m[debug.log] 12: SAVE 5[0m
[32m[debug.log] 13: MATCH[0m
[32m[debug.log] Some(0, \\"aabbb\\")[0m
[32m[debug.log] Some(0, \\"aa\\")[0m
[32m[debug.log] None[0m
[32m[debug.log] Some(2, \\"bbb\\")[0m
[32m[debug.log] 0: SPLIT 3 1[0m
[32m[debug.log] 1: ANY[0m
[32m[debug.log] 2: JMP -2[0m
[32m[debug.log] 3: SAVE 0[0m
[32m[debug.log] 4: SPLIT 1 5[0m
[32m[debug.log] 5: SPLIT 1 3[0m
[32m[debug.log] 6: CHAR a[0m
[32m[debug.log] 7: JMP -2[0m
[32m[debug.log] 8: JMP -4[0m
[32m[debug.log] 9: SAVE 1[0m
[32m[debug.log] 10: MATCH[0m
[32m[debug.log] true[0m
[32m[debug.log] Some(7, \\"le\\")[0m
[32m[debug.log] Some(7, \\"le\\")[0m
[32m[debug.log] Some(10, \\"apple\\")[0m
[32m[debug.log] Some(0, \\"  \\")[0m
[32m[debug.log] ()|((a)())[0m
[32m[debug.log] ()|((.)())[0m
[32m[debug.log] ()|((\\\\(()|((\\\\(()|((a)())\\\\))())\\\\))())[0m
[32m[debug.log] ()|((\\\\(()|(((a)*?)())\\\\))())[0m
[32m[debug.log] ()|((\\\\(()|((a)((\\\\(()|((c)())\\\\))()))\\\\))((\\\\(()|((b)())\\\\))()))[0m
[32m[debug.log] ()|((\\\\((()|(((a)+)()))|((\\\\(()|((c)())\\\\))())\\\\))((\\\\(()|(((b)+)())\\\\))()))[0m
[32m[debug.log] true[0m
[32m[debug.log] ()|((
)((
)((	)(($)((()(())((*)((?)()))))))))[0m
[32m[debug.log] true[0m
[32m[debug.log] true[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] true[0m
[32m[debug.log] true[0m
[32m[debug.log] ==== Parser ====[0m
[32m[debug.log] aaa[0m
[32m[debug.log] 3[0m
[32m[debug.log] 10[0m
[32m[debug.log] bar[0m
[32m[debug.log] ==== Lexer ====[0m
[32m[debug.log] (| kind = identifier, data = \\"abc\\" |)[0m
[32m[debug.log] (| kind = whitespace, data = \\" \\" |)[0m
[32m[debug.log] (| kind = identifier, data = \\"defg\\" |)[0m
[32m[debug.log] (| kind = whitespace, data = \\"   \\" |)[0m
[32m[debug.log] (| kind = identifier, data = \\"hijklMNL\\" |)[0m
[32m[debug.log] (| kind = whitespace, data = \\" \\" |)[0m
[32m[debug.log] (| kind = identifier, data = \\"op123\\" |)[0m
[32m[debug.log] ==== Int ====[0m
[32m[debug.log] 52[0m
[32m[debug.log] 13[0m
[32m[debug.log] 6[0m
[32m[debug.log] 3[0m
[32m[debug.log] 1[0m
[32m[debug.log] 0[0m
[32m[debug.log] 0[0m
[32m[debug.log] -13[0m
[32m[debug.log] -7[0m
[32m[debug.log] -4[0m
[32m[debug.log] -2[0m
[32m[debug.log] -1[0m
[32m[debug.log] -1[0m
[32m[debug.log] 8[0m
[32m[debug.log] 19[0m
[32m[debug.log] 15[0m
[32m[debug.log] 7[0m
[32m[debug.log] 3[0m
[32m[debug.log] 2305843009213693948[0m
[32m[debug.log] 12345678[0m
[32m[debug.log] -123[0m
[32m[debug.log] 65535[0m
[32m[debug.log] 8[0m
[32m[debug.log] 1[0m
[32m[debug.log] 3[0m
[32m[debug.log] -4611686018427387904[0m
[32m[debug.log] Some(123)[0m
[32m[debug.log] Some(-123)[0m
[32m[debug.log] None[0m
[32m[debug.log] Some(10)[0m
[32m[debug.log] ==== Map ====[0m
[32m[debug.log] a[0m
[32m[debug.log] b[0m
[32m[debug.log] c[0m
[32m[debug.log] a[0m
[32m[debug.log] d[0m
[32m[debug.log] c[0m
[32m[debug.log] ==== Promise ====[0m
[32m[debug.log] bar[0m
[32m[debug.log] foo[0m
[32m[debug.log] barbaz[0m
[32m[debug.log] foobar[0m
[32m[debug.log] ==== Float ====[0m
[32m[debug.log] 3.14159265359[0m
[32m[debug.log] 2.71828182846[0m
[32m[debug.log] inf == inf = true[0m
[32m[debug.log] inf == -inf = false[0m
[32m[debug.log] inf == -nan = false[0m
[32m[debug.log] inf == 0. = false[0m
[32m[debug.log] inf == -0. = false[0m
[32m[debug.log] inf == 1. = false[0m
[32m[debug.log] -inf == inf = false[0m
[32m[debug.log] -inf == -inf = true[0m
[32m[debug.log] -inf == -nan = false[0m
[32m[debug.log] -inf == 0. = false[0m
[32m[debug.log] -inf == -0. = false[0m
[32m[debug.log] -inf == 1. = false[0m
[32m[debug.log] -nan == inf = false[0m
[32m[debug.log] -nan == -inf = false[0m
[32m[debug.log] -nan == -nan = false[0m
[32m[debug.log] -nan == 0. = false[0m
[32m[debug.log] -nan == -0. = false[0m
[32m[debug.log] -nan == 1. = false[0m
[32m[debug.log] 0. == inf = false[0m
[32m[debug.log] 0. == -inf = false[0m
[32m[debug.log] 0. == -nan = false[0m
[32m[debug.log] 0. == 0. = true[0m
[32m[debug.log] 0. == -0. = true[0m
[32m[debug.log] 0. == 1. = false[0m
[32m[debug.log] -0. == inf = false[0m
[32m[debug.log] -0. == -inf = false[0m
[32m[debug.log] -0. == -nan = false[0m
[32m[debug.log] -0. == 0. = true[0m
[32m[debug.log] -0. == -0. = true[0m
[32m[debug.log] -0. == 1. = false[0m
[32m[debug.log] 1. == inf = false[0m
[32m[debug.log] 1. == -inf = false[0m
[32m[debug.log] 1. == -nan = false[0m
[32m[debug.log] 1. == 0. = false[0m
[32m[debug.log] 1. == -0. = false[0m
[32m[debug.log] 1. == 1. = true[0m
[32m[debug.log] inf[0m
[32m[debug.log] -inf[0m
[32m[debug.log] -nan[0m
[32m[debug.log] 23.1406926328[0m
[32m[debug.log] 1.14472988585[0m
[32m[debug.log] 36.4621596072[0m
[32m[debug.log] 2.51097226443e+17[0m
[32m[debug.log] 1.41421356237[0m
[32m[debug.log] 0.497149872694[0m
[32m[debug.log] 2.22507385851e-308[0m
[32m[debug.log] 1.11253692925e-308[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] true[0m
[32m[debug.log] true[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] true[0m
[32m[debug.log] true[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] true[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] true[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] true[0m
[32m[debug.log] false[0m
[32m[debug.log] true[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] false[0m
[32m[debug.log] 2.[0m
[32m[debug.log] 4.[0m
[32m[debug.log] 6.[0m
[32m[debug.log] -2.[0m
[32m[debug.log] -4.[0m
[32m[debug.log] -6.[0m
[32m[debug.log] 6.[0m
[32m[debug.log] 2.[0m
[32m[debug.log] 3.[0m
[32m[debug.log] 5.[0m
[32m[debug.log] -3.[0m
[32m[debug.log] -4.[0m
[32m[debug.log] -6.[0m
[32m[debug.log] 6.[0m
[32m[debug.log] 3.[0m
[32m[debug.log] 4.[0m
[32m[debug.log] 6.[0m
[32m[debug.log] -2.[0m
[32m[debug.log] -3.[0m
[32m[debug.log] -5.[0m
[32m[debug.log] 6.[0m
[32m[debug.log] 2.[0m
[32m[debug.log] 3.[0m
[32m[debug.log] 5.[0m
[32m[debug.log] -2.[0m
[32m[debug.log] -3.[0m
[32m[debug.log] -5.[0m
[32m[debug.log] 6.[0m
[32m[debug.log] 9.88131291682e-324[0m
[32m[debug.log] 4.94065645841e-324[0m
[32m[debug.log] 0.[0m
[32m[debug.log] 4.49423283716e+307[0m
[32m[debug.log] 8.98846567431e+307[0m
[32m[debug.log] inf[0m
[32m[debug.log] Some(123.45)[0m
[32m[debug.log] Some(-123.45)[0m
[32m[debug.log] Some(123.)[0m
[32m[debug.log] Some(0.111)[0m
[32m[debug.log] None[0m
[32m[debug.log] ==== Bool ====[0m
[32m[debug.log] true[0m
[32m[debug.log] false[0m
[32m[debug.log] Some(false)[0m
[32m[debug.log] Some(true)[0m
[32m[debug.log] None[0m
[32m[debug.log] Some(false)[0m
[32m[debug.log] Some(true)[0m
[32m[debug.log] None[0m
[32m[debug.log] ==== Stream ====[0m
[32m[debug.log] 1[0m
[32m[debug.log] 1[0m
[32m[debug.log] 2[0m
[32m[debug.log] 3[0m
[32m[debug.log] 5[0m
[32m[debug.log] 8[0m
[32m[debug.log] 13[0m
[32m[debug.log] 21[0m
[32m[debug.log] 34[0m
[32m[debug.log] 55[0m
[32m[debug.log] 89[0m
[32m[debug.log] 144[0m
[32m[debug.log] 233[0m
[32m[debug.log] 377[0m
[32m[debug.log] 610[0m
[32m[debug.log] 987[0m
[32m[debug.log] 1597[0m
[32m[debug.log] 2584[0m
[32m[debug.log] 4181[0m
[32m[debug.log] 6765[0m
[32m[debug.log] 10946[0m
[32m[debug.log] 17711[0m
[32m[debug.log] 28657[0m
[32m[debug.log] 46368[0m
[32m[debug.log] 75025[0m
[32m[debug.log] 121393[0m
[32m[debug.log] 196418[0m
[32m[debug.log] 317811[0m
[32m[debug.log] 514229[0m
[32m[debug.log] 832040[0m
[32m[debug.log] 0[0m
[32m[debug.log] 1[0m
[32m[debug.log] 2[0m
[32m[debug.log] 3[0m
[32m[debug.log] 4[0m
[32m[debug.log] ==== Fn ====[0m
[32m[debug.log] 0[0m
[32m[debug.log] 1[0m
[32m[debug.log] 2[0m
[32m[debug.log] 3[0m
[32m[debug.log] 4[0m
[32m[debug.log] 5[0m
[32m[debug.log] 6[0m
[32m[debug.log] 7[0m
[32m[debug.log] 8[0m
[32m[debug.log] 9[0m
[32m[debug.log] done[0m
"
`;
